### PR TITLE
[FirmwareFlasher] Adds logging

### PR DESCRIPTION
**Pull request by krautech**
## Description

^ as title suggests. Open to thoughts.

### CODE REVIEW

File: scripts/firmware.py [hunk 1]
```diff
@@ -102,6 +104,44 @@ class FirmwareFile(NamedTuple):
...
 class Utils:
+    @staticmethod
+    def configure_logging():
+        # Get the root logger
+        logger = logging.getLogger()
+        logger.setLevel(logging.DEBUG)  # Capture all logs (DEBUG and above)
+
+        # Remove all existing handlers
+        if logger.hasHandlers():
+            logger.handlers.clear()
```
Comment (Jomik) [inline scripts/firmware.py:115]: Is this really necessary?

File: scripts/firmware.py [hunk 2]
```diff
@@ -102,6 +104,44 @@ class FirmwareFile(NamedTuple):
...
+
+        # Create a console handler (always active)
+        console_handler = logging.StreamHandler()
+        console_handler.setLevel(
+            logging.INFO
+        )  # Show only info and above in the console
```
Comment (Jomik) [inline scripts/firmware.py:121]: Will this not mess with the UI?

File: scripts/firmware.py [hunk 3]
```diff
@@ -163,14 +203,22 @@ def colored_text(text: str, color: Color) -> str:
...
+        colored_message = Utils.colored_text(f"Error: {message}", Color.RED)
+        logging.error(colored_message)
+        print(colored_message)
         _ = input(Utils.colored_text("\nPress Enter to continue...", Color.YELLOW))
 
     @staticmethod
     def success_msg(message: str) -> None:
-        print(Utils.colored_text("Success:", Color.GREEN), message)
+        colored_message = Utils.colored_text(f"Success: {message}", Color.GREEN)
+        logging.debug(message)  # Log the colored message
```
Comment (Jomik) [inline scripts/firmware.py:214]: Probably shouldn't have colors in the logging output, that will be _really_ hard to read in the file?
I'd argue that we should use a prefix and then the console logger handler could see that prefix and map it to the color instead.

File: scripts/firmware.py [hunk 4]
```diff
@@ -163,14 +203,22 @@ def colored_text(text: str, color: Color) -> str:
 
     @staticmethod
     def error_msg(message: str) -> None:
-        print(Utils.colored_text("Error:", Color.RED), message)
+        colored_message = Utils.colored_text(f"Error: {message}", Color.RED)
+        logging.error(colored_message)
+        print(colored_message)
```
Comment (Jomik) [inline scripts/firmware.py:208]: Isn't this ending up printing it twice?

File: scripts/firmware.py [hunk 5]
```diff
@@ -163,14 +203,22 @@ def colored_text(text: str, color: Color) -> str:
...
     def success_msg(message: str) -> None:
-        print(Utils.colored_text("Success:", Color.GREEN), message)
+        colored_message = Utils.colored_text(f"Success: {message}", Color.GREEN)
+        logging.debug(message)  # Log the colored message
+        print(colored_message)
         _ = input(Utils.colored_text("\nPress Enter to continue...", Color.YELLOW))
 
+    @staticmethod
+    def log_debug(message: str) -> None:
+        logging.debug(message)
```
Comment (Jomik) [inline scripts/firmware.py:220]:
```suggestion
```
You're not using it  anyway 😄

File: scripts/firmware.py [hunk 6]
```diff
@@ -1444,6 +1502,54 @@ def query_devices(self):
         menu = Menu("Options", menu_items)
         menu.display()
 
+    def enter_katapult_bootloader(self, device: str):
```
Comment (Jomik) [inline scripts/firmware.py:1505]: Is this a fully new function? Should it be separate PR?

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scripts/firmware.py (hunk 2 above)
Comment (krautech) [reply scripts/firmware.py:121]: No? I made sure it wouldnt :)

File: scripts/firmware.py (hunk 1 above)
Comment (krautech) [reply scripts/firmware.py:115]: Removed

File: scripts/firmware.py (hunk 3 above)
Comment (krautech) [reply scripts/firmware.py:214]: If you look, its sending the uncolored message to logging, and colored to print

oops.. i had fixed success but not error. Have now.

File: scripts/firmware.py (hunk 4 above)
Comment (krautech) [reply scripts/firmware.py:208]: Only prints to console `logging.info`

Comment (krautech) [reply scripts/firmware.py:208]:
```
console_handler.setLevel(
   logging.INFO
)  # Show only info and above in the console
        ```

File: scripts/firmware.py (hunk 6 above)
Comment (krautech) [reply scripts/firmware.py:1505]: renamed and changed a little, was needed for logging to be able to capture the outputs as it couldnt before.

Comment (krautech) [reply scripts/firmware.py:1497]: Oh yes I did move it to a seperate function to make it a little clearer. But i cbf with a seperate PR for it since its needed as apart of this PR.

File: scripts/firmware.py (hunk 2 above)
Comment (krautech) [reply scripts/firmware.py:117]: Oh and above.. hmm

File: scripts/firmware.py (hunk 4 above)
Comment (krautech) [reply scripts/firmware.py]: Okay i lied, fixed it

File: scripts/firmware.py (hunk 2 above)
Comment (krautech) [reply scripts/firmware.py]: Fixed with a filter

File: scripts/firmware.py (hunk 3 above)
Comment (Jomik) [reply scripts/firmware.py:216]: Ahh, it was the comment that confused me.

File: scripts/firmware.py (hunk 6 above)
Comment (Jomik) [reply scripts/firmware.py:1503]: Super.

**Review by Jomik — APPROVED** [review]
